jps-plugin/src/org/intellij/erlang/jps/builder/ErlangPrepareDependenciesTarget.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2012-2015 Sergey Ignatov
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.intellij.erlang.jps.builder;
+
+import com.intellij.util.containers.ContainerUtil;
+import org.intellij.erlang.jps.model.ErlangIncludeSourceRootType;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.jps.builders.BuildRootIndex;
+import org.jetbrains.jps.builders.BuildTarget;
+import org.jetbrains.jps.builders.BuildTargetRegistry;
+import org.jetbrains.jps.builders.TargetOutputIndex;
+import org.jetbrains.jps.builders.storage.BuildDataPaths;
+import org.jetbrains.jps.incremental.CompileContext;
+import org.jetbrains.jps.indices.IgnoredFileIndex;
+import org.jetbrains.jps.indices.ModuleExcludeIndex;
+import org.jetbrains.jps.model.JpsDummyElement;
+import org.jetbrains.jps.model.JpsModel;
+import org.jetbrains.jps.model.JpsProject;
+import org.jetbrains.jps.model.java.JavaSourceRootProperties;
+import org.jetbrains.jps.model.java.JavaSourceRootType;
+import org.jetbrains.jps.model.module.JpsModule;
+import org.jetbrains.jps.model.module.JpsTypedModuleSourceRoot;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+public class ErlangPrepareDependenciesTarget extends BuildTarget<ErlangSourceRootDescriptor> {
+  private static final String NAME = "prepare dependencies target";
+  private static ErlangPrepareDependenciesTarget ourInstance;
+
+  private final JpsProject myProject;
+
+  private ErlangPrepareDependenciesTarget(@NotNull JpsProject project,@NotNull ErlangPrepareDependenciesTargetType targetType) {
+    super(targetType);
+    myProject = project;
+  }
+
+  public static ErlangPrepareDependenciesTarget getInstance(@NotNull JpsProject project,
+                                                            @NotNull ErlangPrepareDependenciesTargetType targetType) {
+    if (ourInstance == null) {
+      ourInstance = new ErlangPrepareDependenciesTarget(project, targetType);
+    }
+    return ourInstance;
```

Comment:
What's the purpose of this shared instance? Why isn't it thread-safe? How come, you're sure your target will only be instantiated for a single project during your JVM's session?